U/wmwv/dr6 - Add script and Notebook for validation of Run 2.2i DR6a

wmwv opened the pull request:
1. Notebook for visual interaction.  Can use either direct DPDD parquet file or GCRCatalogs API.  Memory performance is an issue if you try to load the full dataset using the JupyterHub at NERSC.
2. Script for producing plots and for running on a dedicated node with enough memory to process full dataset.  Based on DPDD Parquet file, with a correction to recompute psf_fwhm.

This is the script that produced the plots used for the DR6a validation I presented at
https://confluence.slac.stanford.edu/x/mYDPE

Fixes #119

yymao commented on validation/validate_dc2_run2.2i_object_table.ipynb:
```diff
@@ -0,0 +1,1367 @@
+   "source": [
+    "sampling_factor = 100\n",
+    "restrict_to_tracts = None\n",
+    "# tracts = [3640]"
```
```suggestion
    "# restrict_to_tracts = [3640]"
```

wmwv replied:
Thank you for the catch.

yymao commented on validation/validate_dc2_run2.2i_object_table.ipynb:
```diff
@@ -0,0 +1,1367 @@
+    "print(f'Total: {len(df)}, Good: {len(good_idx)}, Stars: {len(star_idx)}, Galaxies: {len(galaxy_idx)}')\n",
+    "if read_from_dpdd_parquet:\n",
+    "    print(f'For {catalog_file} with {sampling_factor}x subsample')\n",
+    "    if tracts:\n",
+    "        print(f'Tracts: ', tracts)\n",
+    "else:\n",
+    "    print(f'For {catalog_name} with {sampling_factor}x subsample')\n",
+    "    if tracts:\n",
+    "        print(f'Tracts: ', tracts)"
```
`tracts` -> `restrict_to_tracts`

yymao commented on validation/validate_dc2_run2.2i_object_table.ipynb:
```diff
@@ -0,0 +1,1367 @@
+    "fig, axes = plt.subplots(2, 3, figsize=(12, 6))\n",
+    "for ax, filt in zip(axes.flat, filters):\n",
+    "    mag_col, magerr_col = f'mag_{filt}', f'magerr_{filt}'\n",
+    "    plot_mag_magerr(df.loc[galaxy_idx, [mag_col, magerr_col]], filt, ax=ax)\n",
```
`vmin` argument for `plot_mag_magerr` is default to 100, which may or may not work depending on how many tracts one loads. Add a comments about changing `vmin` is an error occurs.

yymao commented on validation/validate_dc2_run2.2i_object_table.ipynb:
```diff
@@ -0,0 +1,1367 @@
+    "fig, axes = plt.subplots(2, 3, figsize=(12, 6))\n",
+    "for ax, filt in zip(axes.flat, filters):\n",
+    "    mag_col, magerr_col = f'mag_{filt}', f'magerr_{filt}'\n",
+    "    plot_mag_magerr(df.loc[star_idx, [mag_col, magerr_col]], filt, ax=ax)\n",
```
`vmin` argument for `plot_mag_magerr` is default to 100, which may or may not work depending on how many tracts one loads. Add a comments about changing `vmin` is an error occurs.

yymao commented on validation/validate_dc2_run2.2i_object_table.ipynb:
```diff
@@ -0,0 +1,1367 @@
+    "legend = True\n",
+    "for ax, filt in zip(axes.flat, filters):\n",
+    "    plot_shape(df, filt, good_idx, star_idx, galaxy_idx, ax=ax, legend=legend)\n",
+    "    legend = False"
```
After the for loop, add `plt.tight_layout()` so that the plot looks nicer (less overlapping text)

yymao commented on validation/validate_dc2_run2.2i_object_table.ipynb:
```diff
@@ -0,0 +1,1367 @@
+    "legend = True\n",
+    "for ax, filt in zip(axes.flat, filters):\n",
+    "    plot_ellipticity(df, filt, good_idx, star_idx, galaxy_idx, ax=ax, legend=legend)\n",
+    "    legend = False"
```
After the for loop, add `plt.tight_layout()` so that the plot looks nicer (less overlapping text)

yymao commented on validation/validate_dc2_run2.2i_object_table.py:
```diff
@@ -0,0 +1,870 @@
+
+
+def print_expected_memory_usage(columns, sampling_factor):
+    N = 52000000
+    MB_per_column = 512 * (N / 64000000)  # MB / column / 64 million rows
+    print(f"We are going to load {len(columns)} columns.")
```
See my comment above.

yymao commented on validation/validate_dc2_run2.2i_object_table.py:
```diff
@@ -0,0 +1,870 @@
+    data_release = "DC2_Run2.2i_DR6a"
+    sampling_factor = 1
+
+    filters, df = load_data(sampling_factor=sampling_factor)
```
Here one should pass the `data_release` variable to `load_data` otherwise it leaves impression that `load_data` can magically know which data release to use.

wmwv commented on validation/validate_dc2_run2.2i_object_table.ipynb:
```diff
@@ -0,0 +1,1367 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "N = 52000000\n",
+    "MB_per_column = 512 * (N / 64000000) # MB / column / 64 million rows\n",
+    "print(f'We are going to load {len(columns)} columns.')\n",
+    "print(f'For {N//sampling_factor} million rows that should take {(len(columns)/sampling_factor)*MB_per_column/1024:0.2f} GB of memory')"
```
Ah, yes.  512 MB / column was the reference number for 64 million rows.
I've cleaned up the variable naming and logic a bit.  Thanks.

wmwv commented on validation/validate_dc2_run2.2i_object_table.ipynb:
```diff
@@ -0,0 +1,1367 @@
+   "outputs": [],
+   "source": [
+    "if sampling_factor > 1:\n",
+    "    arbitrary_last_digit = '2'  # If you're really bored, you can explore other choices for the last digit sequence.\n",
+    "    remainder = int(math.ceil(math.log10(sampling_factor)) * arbitrary_last_digit)\n",
+    "    # You can use sampling_factors that are not a power of ten.\n",
+    "    # If you do so, we need to make sure that the remainder is less than the modulus\n",
+    "    remainder = remainder % sampling_factor\n",
```
Because if the user chose

wmwv replied:
Wow, I really overthought that one.  I can now recreate my thinking about it, but we will both be happier if I don't try to explain.

wmwv commented:
1. Removed special sys.path that's no longer required.
2. Reworked subsampling calculation
3. Added a bit of logic in the load_data (will rework in the future to hardcode less.)
4. Tested both script and Notebook.

Ready for approval.
